ipywidgets/static/widgets/js/manager.js

```diff
@@ -268,6 +288,96 @@ define([
         // exists).
         return null;
     };
+    
+    /**
+     * Updates rendered snapshots of all of the widget areas in the notebook.
+     * @return {Promise<void>} success
+     */
+    WidgetManager.prototype.updateSnapshots = function() {
+        var cells = Jupyter.notebook.get_cells();
+        return Promise.all(cells.map((function(cell) {
+            var widgetSubarea = cell.element[0].querySelector(".widget-subarea");
+            if (widgetSubarea.children.length > 0) {
+                return new Promise((function(resolve) {
+                    html2canvas(widgetSubarea, {
+                        onrendered: function(canvas) {
+                            // Save the screenshot of the canvas to the widget-subarea
+                            var mimetype = "image/png";
+                            widgetSubarea.widgetSnapshot = {
+                                mimetype: mimetype,
+                                data: canvas.toDataURL(mimetype)
+                            };
+                            
+                            resolve();
+                        }
+                    });
+                }).bind(this));
+            } else {
+                if (widgetSubarea.widgetSnapshot) {
+                    delete widgetSubarea.widgetSnapshot;
+                }
+            }
+        }).bind(this)));
+    };
+    
+    /**
+     * Render the widget views that are live as images and prepend them as
+     * outputs.
+     *
+     * Note: This function must be synchronous, in order to work with the 
+     * notebook's save machinery.
+     */
+    WidgetManager.prototype.prependSnapshots = function() {
+        var cells = Jupyter.notebook.get_cells();
+        cells.forEach((function(cell) {
+            var widgetSubarea = cell.element[0].querySelector(".widget-subarea");
+            if (widgetSubarea.children.length > 0 && widgetSubarea.widgetSnapshot) {
+                
+                // Get the last screenshot of the widget sub-area
+                var mimetype = widgetSubarea.widgetSnapshot.mimetype;
+                var screenshot = widgetSubarea.widgetSnapshot.data;
+                
+                // Create a mime bundle for the screenshot. Remove
+                // URL information, so only the b64 encoded data 
+                // exists, because that's what the notebook likes.
```

Comment:
and we should continue to question this assumption in future versions of both the message spec and the notebook